### PR TITLE
In/Out-columns in DashAIDataset

### PR DESCRIPTION
**Pull request by MillarayVVS**
# Summary

Datasets Refactor deleting all the usage of input and output columns on other components.

## Type of change

- Refactoring.

## Changes

Remove use of input and output columns on DashAIDataset class:
- Delete getters and setters.
- Modify save and load fuctions, now they don't include input and output columns.
- Modify validation of input and output columns, now they are passed as parameters to the method.

With the modifications above it was necessary to modify all the usage of DashAIDataset on other components:

Task:
- Modify prepare and validate dataset methods, now the input and output columns are passed as parameters.

DataLoader:
- Delete input and output columns to create a new DashAIDataset.

Metrics:
- Adds parameter with output columns to use in prepare to metric function.

Model:
- Refactor of fit function, now the dataset is divided in two on Model Job so this method receive a dataset with the input columns and a dataset with the output columns as two different parameters.
- Same with predict function, now this method only receive the dataset with the output columns.

image

All this changes include the tests refactoring.

## How to Test

1.- Run ```pytest tests/```

### CODE REVIEW

File: DashAI/back/dataloaders/classes/dashai_dataset.py [hunk 1]
```diff
@@ -249,45 +163,6 @@ def sample(
...
-    if set(inputs + outputs) != set(names):
-        raise ValueError(
-            "The union of the elements of inputs and outputs list must be equal to "
-            "elements in the list of names."
-        )
-
-
 @beartype
 def load_dataset(dataset_path: str) -> DatasetDict:
     """Load a datasetdict with dashaidatasets inside.
```
Comment (pbadillatorrealba) [inline DashAI/back/dataloaders/classes/dashai_dataset.py]:
```suggestion
    """Load a DashAI dataset from its path.

         This process cast each split into a DashAIdataset object.
```

File: DashAI/back/dataloaders/classes/dashai_dataset.py [hunk 2]
```diff
@@ -352,6 +216,81 @@ def save_dataset(datasetdict: DatasetDict, path: str) -> None:
...
 
 
+@beartype
+def validate_inputs_outputs(
+    datasetdict: DatasetDict,
+    inputs: List[str],
+    outputs: List[str],
+) -> None:
+    """Validate the columns to be chosen as input and output.
+    The algorithm considers those that already exist in the dataset.
```
Comment (pbadillatorrealba) [inline DashAI/back/dataloaders/classes/dashai_dataset.py:227]: Add a new line between the first line and the next paragraph in a docstring.

File: DashAI/back/dataloaders/classes/dashai_dataset.py [hunk 3]
```diff
@@ -352,6 +216,81 @@ def save_dataset(datasetdict: DatasetDict, path: str) -> None:
...
+            "Inputs and outputs cannot have more elements than names. "
+            f"Number of inputs: {len(inputs)}, "
+            f"number of outputs: {len(outputs)}, "
+            f"number of names: {len(dataset_features)}. "
+        )
+        # Validate that inputs and outputs only contain elements that exist in names
+    if not set(dataset_features).issuperset(set(inputs + outputs)):
+        raise ValueError(
+            "Inputs and outputs can only contain elements that exist in names."
+        )
```
Comment (pbadillatorrealba) [inline DashAI/back/dataloaders/classes/dashai_dataset.py]: I think it is a good idea to show the extra features detected:
```suggestion
        raise ValueError(
            "Inputs and outputs can only contain elements that exist in names. Extra elements: %s",
            str(set(input+output). difference(set(datasets_features))) # check this op
        )
```

File: DashAI/back/dataloaders/classes/dashai_dataset.py [hunk 4]
```diff
@@ -352,6 +216,81 @@ def save_dataset(datasetdict: DatasetDict, path: str) -> None:
         )
 
 
+@beartype
+def validate_inputs_outputs(
+    datasetdict: DatasetDict,
+    inputs: List[str],
+    outputs: List[str],
+) -> None:
+    """Validate the columns to be chosen as input and output.
```
Comment (pbadillatorrealba) [inline DashAI/back/dataloaders/classes/dashai_dataset.py:226]: Nothing happens if the lists are empty?

File: DashAI/back/dataloaders/classes/dashai_dataset.py [hunk 5]
```diff
@@ -352,6 +216,81 @@ def save_dataset(datasetdict: DatasetDict, path: str) -> None:
...
+        )
+        # Validate that inputs and outputs only contain elements that exist in names
+    if not set(dataset_features).issuperset(set(inputs + outputs)):
+        raise ValueError(
+            "Inputs and outputs can only contain elements that exist in names."
+        )
+
+
+@beartype
+def parse_columns_indices(dataset_path: str, indices: List[int]) -> List[str]:
```
Comment (pbadillatorrealba) [inline DashAI/back/dataloaders/classes/dashai_dataset.py:260]: Add documentation.

File: DashAI/back/dataloaders/classes/dashai_dataset.py [hunk 6]
```diff
@@ -352,6 +216,81 @@ def save_dataset(datasetdict: DatasetDict, path: str) -> None:
...
+    dataset = load_dataset(dataset_path=dataset_path)
+    dataset_features = list((dataset["train"].features).keys())
+    names_list = []
+    for index in indices:
```
Comment (pbadillatorrealba) [inline DashAI/back/dataloaders/classes/dashai_dataset.py:273]: `indices` comes from the front?
Maybe its a good idea to preprocess it outside this function, maybe in the same request code in the API.

File: DashAI/back/dataloaders/classes/dashai_dataset.py [hunk 7]
```diff
@@ -352,6 +216,81 @@ def save_dataset(datasetdict: DatasetDict, path: str) -> None:
...
+        if index <= len(dataset_features):
```
Comment (pbadillatorrealba) [inline DashAI/back/dataloaders/classes/dashai_dataset.py]: If the validation does not fail, this should not happen right?

Anyway, I suggest adding an else where you throw an exception if this case occurs.

File: DashAI/back/dataloaders/classes/dashai_dataset.py [hunk 8]
```diff
@@ -352,6 +216,81 @@ def save_dataset(datasetdict: DatasetDict, path: str) -> None:
...
+    output_columns : List[str]
+        List with the output columns labels
+
+    Returns
+    -------
+    Dict
+        Dict with the splits divided in x and y tuple
+    """
+    input_columns_dataset = copy.deepcopy(dataset)
+    output_columns_dataset = copy.deepcopy(dataset)
```
Comment (pbadillatorrealba) [inline DashAI/back/dataloaders/classes/dashai_dataset.py]: I think it would be a good idea to see if the original dataset object implements copy and use that.
It could be that, making an efficient memory management, this object does not contain the dataset, but only a reference to it and that at the moment of deepcopy, only the reference is copied and not the complete dataset.

File: DashAI/back/dataloaders/classes/dashai_dataset.py [hunk 9]
```diff
@@ -352,6 +216,81 @@ def save_dataset(datasetdict: DatasetDict, path: str) -> None:
...
+@beartype
+def select_columns(
+    dataset: DatasetDict, input_columns: List[str], output_columns: List[str]
+) -> Tuple[DatasetDict, DatasetDict]:
+    """Load and prepare the dataset into dataframes to use in models.
+
+    Parameters
+    ----------
+    dataset : DashAIDataset
+        Dataset to format
```
Comment (pbadillatorrealba) [inline DashAI/back/dataloaders/classes/dashai_dataset.py]: Update the docstring.

File: DashAI/back/dataloaders/classes/dashai_dataset.py (hunk 8 above)
Comment (pbadillatorrealba) [reply DashAI/back/dataloaders/classes/dashai_dataset.py]: Also, `select_colums` will not make copies?

File: DashAI/back/dataloaders/classes/dashai_dataset.py [hunk 10]
```diff
@@ -352,6 +216,81 @@ def save_dataset(datasetdict: DatasetDict, path: str) -> None:
...
+    input_columns : List[str]
+        List with the input columns labels
+    output_columns : List[str]
+        List with the output columns labels
+
+    Returns
+    -------
+    Dict
+        Dict with the splits divided in x and y tuple
```
Comment (pbadillatorrealba) [inline DashAI/back/dataloaders/classes/dashai_dataset.py]: Update output docs

File: DashAI/back/api/api_v1/endpoints/predict.py [hunk 11]
```diff
@@ -110,17 +102,19 @@ async def perform_predict(
...
-        filepath_or_buffer=input_file, temp_path=tmp_path, params={"data_key": "data"}
+        filepath_or_buffer=input_file,
+        temp_path=tmp_path,
+        params={"data_key": "data"},
     )
+    # TODO Extract this Code to DashAIDataset
     input_df = pd.DataFrame(raw_dataset["train"])
-    # TODO: Use feature_names from Experiment
-    input_df = input_df.reindex(columns=json.loads(dat.feature_names))
+    input_df = input_df.reindex(columns=exp.input_columns)
```
Comment (pbadillatorrealba) [inline DashAI/back/api/api_v1/endpoints/predict.py:114]: Does this functionality was not done in `select_columns`?

File: DashAI/back/dataloaders/classes/dashai_dataset.py
Comment (pbadillatorrealba) [inline DashAI/back/dataloaders/classes/dashai_dataset.py:1]: Since the columns are now handled externally in the experiment, does it make sense to continue building a wrapper?
Why not just use Datasets directly?

File: DashAI/back/dataloaders/classes/dataloader.py [hunk 12]
```diff
@@ -92,22 +92,10 @@ def extract_files(self, dataset_path: str, file: UploadFile) -> str:
 
     def _check_split_values(
         self,
-        dataset: DatasetDict,
         train_size: float,
         test_size: float,
         val_size: float,
-        stratify: bool = False,
```
Comment (pbadillatorrealba) [inline DashAI/back/dataloaders/classes/dataloader.py:99]: Will stratify no longer be implemented?

File: DashAI/back/models/scikit_learn/sklearn_like_model.py [hunk 13]
```diff
@@ -22,41 +21,24 @@ def load(filename: str) -> None:
...
-        x, y = self.format_data(dataset)
-        return super().fit(x, y)
+        x_pandas = x.to_pandas()
+        y_pandas = y.to_pandas()
+        return super().fit(x_pandas, y_pandas)
 
-    def predict(self, dataset: DashAIDataset):
-        x, y = self.format_data(dataset)
-        return super().predict_proba(x)
+    def predict(self, x: DashAIDataset):
```
Comment (pbadillatorrealba) [inline DashAI/back/models/scikit_learn/sklearn_like_model.py]: Document this function

File: DashAI/back/models/scikit_learn/sklearn_like_model.py [hunk 14]
```diff
@@ -22,41 +21,24 @@ def load(filename: str) -> None:
...
         """Fit the estimator.
 
         Parameters
         ----------
-        dataset : DashAIDataset
-            The training dataset.
+        x : pd.DataFrame
+            Dataframe with the input data.
+        y : pd.DataFrame
+            Dataframe with the output data.
```
Comment (pbadillatorrealba) [inline DashAI/back/models/scikit_learn/sklearn_like_model.py]: Why not to use `X_train` and `y_train` here , and in `X_pred` in the prediction function?

**Review by pbadillatorrealba — CHANGES_REQUESTED** [review]
I left comments on the code of the changes, I am still pending with the review of the tests.

File: DashAI/back/dataloaders/classes/dashai_dataset.py (hunk 4 above)
Comment (MillarayVVS) [reply DashAI/back/dataloaders/classes/dashai_dataset.py:226]: I added an Error to throw when this happens

File: DashAI/back/api/api_v1/endpoints/predict.py (hunk 11 above)
Comment (Riul1999) [reply DashAI/back/api/api_v1/endpoints/predict.py:114]: No, the above code reorders the columns in the dataset to match the column order when the model was trained. The `select_columns` method of **DashAIDataset** selects a subset of the actual columns in the **Dataset**, but does not affect the column order.

File: DashAI/back/dataloaders/classes/dataloader.py (hunk 12 above)
Comment (Riul1999) [reply DashAI/back/dataloaders/classes/dataloader.py:99]: We are going to move the _split_dataset_ method from the **Dataloader** class to the **DashAIDataset** class, because the splitting of the dataset is done at experiment creation (linked to the **DashAIDataset**) instead of dataset creation (linked to the **Dataloader**). So we will add the option to _stratify_ in this new method.

**Review by pbadillatorrealba — APPROVED** [review]
Looks good to me!
